ARM64环境下，使用gaussdb 506.0SPC0100版本的libpq,在连接失败的时候python进程会coredump

```
(.env311) [root@ae910d8786f844c4b317045140860809 test_gaussdb]# cat test1.py 
#!/usr/bin/env python3
import argparse
import os
import sys

def main() -> int:

    try:
        import gaussdb
    except Exception as exc:
        print(f"[ERROR] Failed to import gaussdb: {exc}")
        return 3

    print("[INFO] gaussdb module:", getattr(gaussdb, "__file__", "<unknown>"))
    print("[INFO] gaussdb version:", getattr(gaussdb, "__version__", "<unknown>"))

    conn = None
    try:
        conn = gaussdb.connect('host=127.0.0.1 port=3333 dbname=database_name user=username password=password', connect_timeout=10)
        with conn.cursor() as cur:
            cur.execute("select 1")
            row = cur.fetchone()
        print("[OK] Connection succeeded, query result:", row)
        return 0
    except Exception as exc:
        print(f"[ERROR] Connection/query failed: {exc}")
        return 1
    finally:
        if conn is not None:
            try:
                conn.close()
            except Exception:
                pass


if __name__ == "__main__":
    sys.exit(main())
(.env311) [root@ae910d8786f844c4b317045140860809 test_gaussdb]# export LD_LIBRARY_PATH=/workspace/test_gaussdb/gaussdb-505.2-libpq/lib #使用505.2版本libpq不coredump
(.env311) [root@ae910d8786f844c4b317045140860809 test_gaussdb]# python test1.py 
[INFO] gaussdb module: /workspace/test_gaussdb/gaussdb-python/gaussdb/gaussdb/__init__.py
[INFO] gaussdb version: 1.0.4
[ERROR] Connection/query failed: connection failed: could not connect to server: Operation now in progress
        Is the server running on host "127.0.0.1" and accepting
        TCP/IP connections on port 3333?
(.env311) [root@ae910d8786f844c4b317045140860809 test_gaussdb]# export LD_LIBRARY_PATH=/workspace/test_gaussdb/gaussdb-506.0-libpq/lib  #使用506.0版本libpq会coredump
(.env311) [root@ae910d8786f844c4b317045140860809 test_gaussdb]# python test1.py 
[INFO] gaussdb module: /workspace/test_gaussdb/gaussdb-python/gaussdb/gaussdb/__init__.py
[INFO] gaussdb version: 1.0.4
Segmentation fault (core dumped)
(.env311) [root@ae910d8786f844c4b317045140860809 test_gaussdb]# arch
aarch64
(.env311) [root@ae910d8786f844c4b317045140860809 test_gaussdb]# python --version
Python 3.11.14
```
写C程序直接调用506.0的libpq不会coredump
```
(.env311) [root@ae910d8786f844c4b317045140860809 test_gaussdb]# cat test_pq.c
#include <stdio.h>
#include <stdlib.h>
#include <libpq-fe.h>

int main() {
    // 连接字符串，请根据实际情况修改数据库名、用户名、密码和主机
    PGconn *conn = PQconnectdb("dbname=postgres user=postgres password=secret host=localhost");

    if (PQstatus(conn) != CONNECTION_OK) {
        fprintf(stderr, "连接失败: %s", PQerrorMessage(conn));
        PQfinish(conn);
        exit(1);
    }

    printf("成功连接到 PostgreSQL\n");
    PQfinish(conn);
    return 0;
}(.env311) [root@ae910d8786f844c4b317045140860809 test_gaussdb]# gcc -o test_pq test_pq.c -I/workspace/test_gaussdb/gaussdb-506.0-libpq/include -L/workspace/test_gaussdb/gaussdb-506.0-libpq/lib -lpq
(.env311) [root@ae910d8786f844c4b317045140860809 test_gaussdb]# ./test_pq 
连接失败: could not connect to server: Operation now in progress
        Is the server running on host "localhost" (::1) and accepting
        TCP/IP connections on port 5432?
could not connect to server: Operation now in progress
        Is the server running on host "localhost" (127.0.0.1) and accepting
        TCP/IP connections on port 5432?
(.env311) [root@ae910d8786f844c4b317045140860809 test_gaussdb]# 
```

下面是GPT-5.3-CODEX的分析
>**细化分析**
>- 崩溃链路是：连接失败后在 generators.py 调用 `finish_pgconn()`，随后在 errors.py 读取 `pgconn` 属性快照。
>- 这些属性（如 `host`、`password`）会走到 pq_ctypes.py，内部通过 `_call_bytes()` 调 C API（`PQhost/PQpass`），见 pq_ctypes.py。
>- `_call_bytes()` 对 `c_char_p` 返回值会触发 ctypes 解引用；如果 C 层返回的是“非空但无效地址”，Python 来不及抛异常就会直接 SIGSEGV。你看到的 core 正是这一类崩溃。
>- 我做了独立最小复现：在 `connect_poll` 失败后直接 `print(conn.host)`，无需上层框架也会立刻段错误，所以问题不在异常对象构造本身，而在“失败态 PGconn + PQhost/PQpass 访问”这个组合。
>- 你 C 程序不崩是因为只调用了 `PQstatus/PQerrorMessage/PQfinish`，没有触发 `PQhost/PQpass` 路径。
>
>**为什么说是 libpq 组合问题**
>- 该驱动在 misc.py 用 `find_library("pq")` 动态找库，实际加载结果受 `LD_LIBRARY_PATH`、系统库优先级影响。
>- 你当前这套库报告 `PQlibVersion = 90204 (9.2.4)`，但又来自 gaussdb-506.0-libpq 打包目录，说明是“Gauss 发行库 + 当前系统运行时”的特定组合；在这个组合里，失败态访问某些字符串 getter 不安全。
>
>**结论边界**
>- 我能 100% 确认“崩溃触发点”是 `PQhost/PQpass` 在失败态 `PGconn` 上被调用。
>- 我不能在无 libpq 源码/符号的前提下 100% 断言是 libpq 内部 bug 还是 ABI 兼容性问题，但工程上两者都应按“不安全调用路径”处理

--- a/docs/news.rst
+++ b/docs/news.rst
@@ -11,3 +11,6 @@ gaussdb.0b1
 ^^^^^^^^^^^^^
 
 - First public release on PyPI.
+- Fixed a crash on ARM64 with some libpq builds when connection attempts fail:
+    failed-connection diagnostics now use a safe snapshot instead of reading
+    every libpq connection attribute.

--- a/gaussdb/gaussdb/errors.py
+++ b/gaussdb/gaussdb/errors.py
@@ -22,7 +22,7 @@ from __future__ import annotations
 
 from typing import TYPE_CHECKING, Any, Callable, NoReturn
 from asyncio import CancelledError
-from dataclasses import dataclass, field, fields
+from dataclasses import dataclass, field
 from collections.abc import Sequence
 
 from .pq.abc import PGconn, PGresult
@@ -229,13 +229,42 @@ class FinishedPGconn:
         self._raise()
 
 
-def finish_pgconn(pgconn: PGconn) -> PGconn:
-    args = {}
-    for f in fields(FinishedPGconn):
-        try:
-            args[f.name] = getattr(pgconn, f.name)
-        except Exception:
-            pass
+def finish_pgconn(
+    pgconn: PGconn,
+    *,
+    db: bytes | str | int | None = None,
+    user: bytes | str | int | None = None,
+    host: bytes | str | int | None = None,
+    hostaddr: bytes | str | int | None = None,
+    port: bytes | str | int | None = None,
+    options: bytes | str | int | None = None,
+    error_message: bytes | str | int | None = None,
+    needs_password: bool | None = None,
+) -> PGconn:
+    def _tobytes(value: bytes | str | int | None) -> bytes | None:
+        if value is None:
+            return None
+        if isinstance(value, bytes):
+            return value
+        return str(value).encode("utf-8", "replace")
+
+    args: dict[str, Any] = {}
+    for name, value in (
+        ("db", db),
+        ("user", user),
+        ("host", host),
+        ("hostaddr", hostaddr),
+        ("port", port),
+        ("options", options),
+        ("error_message", error_message),
+    ):
+        bvalue = _tobytes(value)
+        if bvalue is not None:
+            args[name] = bvalue
+
+    if needs_password is not None:
+        args["needs_password"] = bool(needs_password)
+
     pgconn.finish()
     return FinishedPGconn(**args)
 

--- a/gaussdb/gaussdb/generators.py
+++ b/gaussdb/gaussdb/generators.py
@@ -32,6 +32,7 @@ from .abc import Buffer, PipelineCommand, PQGen, PQGenConn
 from .pq.abc import PGcancelConn, PGconn, PGresult
 from .waiting import Ready, Wait
 from ._cmodule import _gaussdb
+from .conninfo import conninfo_to_dict
 from ._encodings import conninfo_encoding
 
 OK = pq.ConnStatus.OK
@@ -69,9 +70,24 @@ def _connect(conninfo: str, *, timeout: float = 0.0) -> PQGenConn[PGconn]:
     while True:
         if conn.status == BAD:
             encoding = conninfo_encoding(conninfo)
+            message = conn.get_error_message(encoding)
+            try:
+                conn_params = conninfo_to_dict(conninfo)
+            except Exception:
+                conn_params = {}
             raise e.OperationalError(
-                f"connection is bad: {conn.get_error_message(encoding)}",
-                pgconn=conn,
+                f"connection is bad: {message}",
+                pgconn=e.finish_pgconn(
+                    conn,
+                    db=conn_params.get("dbname"),
+                    user=conn_params.get("user"),
+                    host=conn_params.get("host"),
+                    hostaddr=conn_params.get("hostaddr"),
+                    port=conn_params.get("port"),
+                    options=conn_params.get("options"),
+                    error_message=message,
+                    needs_password="password" in message.lower(),
+                ),
             )
 
         status = conn.connect_poll()
@@ -89,13 +105,41 @@ def _connect(conninfo: str, *, timeout: float = 0.0) -> PQGenConn[PGconn]:
             break
         elif status == POLL_FAILED:
             encoding = conninfo_encoding(conninfo)
+            message = conn.get_error_message(encoding)
+            try:
+                conn_params = conninfo_to_dict(conninfo)
+            except Exception:
+                conn_params = {}
             raise e.OperationalError(
-                f"connection failed: {conn.get_error_message(encoding)}",
-                pgconn=e.finish_pgconn(conn),
+                f"connection failed: {message}",
+                pgconn=e.finish_pgconn(
+                    conn,
+                    db=conn_params.get("dbname"),
+                    user=conn_params.get("user"),
+                    host=conn_params.get("host"),
+                    hostaddr=conn_params.get("hostaddr"),
+                    port=conn_params.get("port"),
+                    options=conn_params.get("options"),
+                    error_message=message,
+                    needs_password="password" in message.lower(),
+                ),
             )
         else:
+            try:
+                conn_params = conninfo_to_dict(conninfo)
+            except Exception:
+                conn_params = {}
             raise e.InternalError(
-                f"unexpected poll status: {status}", pgconn=e.finish_pgconn(conn)
+                f"unexpected poll status: {status}",
+                pgconn=e.finish_pgconn(
+                    conn,
+                    db=conn_params.get("dbname"),
+                    user=conn_params.get("user"),
+                    host=conn_params.get("host"),
+                    hostaddr=conn_params.get("hostaddr"),
+                    port=conn_params.get("port"),
+                    options=conn_params.get("options"),
+                ),
             )
 
     conn.nonblocking = 1

--- a/tests/test_errors.py
+++ b/tests/test_errors.py
@@ -3,6 +3,7 @@ from __future__ import annotations
 import re
 import sys
 import pickle
+from typing import cast
 from weakref import ref
 
 import pytest
@@ -25,6 +26,36 @@ def test_finishedpgconn(pgconn):
         finished.socket
     with pytest.raises(e.OperationalError, match="connection is closed"):
         pgconn.socket
+
+
+def test_finish_pgconn_doesnt_read_pgconn_attrs():
+    class DummyPGconn:
+        def __init__(self):
+            self.finished = False
+
+        def finish(self):
+            self.finished = True
+
+        @property
+        def db(self):
+            raise AssertionError("db should not be read")
+
+        @property
+        def error_message(self):
+            raise AssertionError("error_message should not be read")
+
+    dummy = DummyPGconn()
+    finished = e.finish_pgconn(
+        cast(pq.abc.PGconn, dummy),
+        db="nosuchdb",
+        error_message="failed",
+        needs_password=True,
+    )
+
+    assert dummy.finished
+    assert finished.db == b"nosuchdb"
+    assert finished.error_message == b"failed"
+    assert finished.needs_password
 
 
 @pytest.mark.crdb_skip("severity_nonlocalized")

--- a/tests/test_generators.py
+++ b/tests/test_generators.py
@@ -35,7 +35,7 @@ def test_connect_operationalerror_pgconn(generators, dsn, monkeypatch):
     pgconn = excinfo.value.pgconn
     assert pgconn is not None
     assert pgconn.needs_password
-    assert b"ERROR:  Invalid username/password,login denied.\n" in pgconn.error_message
+    assert b"Invalid username/password,login denied." in pgconn.error_message
     assert pgconn.status == pq.ConnStatus.BAD.value
     assert pgconn.transaction_status == pq.TransactionStatus.UNKNOWN.value
     assert pgconn.pipeline_status == pq.PipelineStatus.OFF.value

